File: src/main/java/org/modelingvalue/nelumbo/strings/ToInteger.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2026 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo.strings;

import java.io.Serial;
import java.math.BigInteger;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.AstElement;
import org.modelingvalue.nelumbo.NelumboConstructor;
import org.modelingvalue.nelumbo.Node;
import org.modelingvalue.nelumbo.lang.FunctorOrType;
import org.modelingvalue.nelumbo.logic.InferContext;
import org.modelingvalue.nelumbo.logic.InferResult;
import org.modelingvalue.nelumbo.logic.Predicate;

public final class ToInteger extends Predicate {
    @Serial
    private static final long serialVersionUID = -2874326869672600959L;

    @NelumboConstructor
    public ToInteger(FunctorOrType functorOrType, List<AstElement> elements, Node declaration, Object... args) {
        super(functorOrType, elements, declaration, args);
    }

    @Override
    protected ToInteger set(FunctorOrType functorOrType, List<AstElement> elements, Node declaration, Object[] args) {
        return new ToInteger(functorOrType, elements, declaration, args);
    }

    @Override
    protected InferResult infer(int nrOfUnbound, InferContext context) {
        if (nrOfUnbound > 1) {
            return unresolvable();
        }

        BigInteger integer = getVal(0, 0);
        String string = getVal(1, 0);
        if (string != null) {
            try {
                BigInteger parsed = BigInteger.valueOf(Integer.parseInt(string));
                if (integer != null) {
                    boolean eq = integer.equals(parsed);
                    return eq ? factCC() : falsehoodCC();
                } else {
                    return set(0, org.modelingvalue.nelumbo.integers.NInteger.of(parsed)).factCI();
                }
            } catch (NumberFormatException e) {
                return integer != null ? falsehoodCC() : falsehoodCI();
            }
        } else if (integer != null) {
            String s = integer.toString();
            return set(1, org.modelingvalue.nelumbo.strings.NString.of(s)).factCI();
        }

        return unknown();
    }

}

File: web-ts/src/strings/ToInteger.ts
/**
 * ToInteger - String to integer conversion predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.strings.ToInteger
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Predicate } from '../logic/Predicate';
import type { Functor } from '../patterns/Functor';
import type { InferContext } from '../InferContext';
import type { InferResult } from '../InferResult';
import { NInteger } from '../integers/NInteger';
import { NString } from './NString';

export class ToInteger extends Predicate {
  constructor(functor: Functor, elements: List<AstElement>, ...args: unknown[]) {
    super(functor, elements, args[0], args[1]);
  }

  protected static fromDataToInteger(data: unknown[], declaration?: Predicate): ToInteger {
    const node = Object.create(ToInteger.prototype) as ToInteger;
    (node as unknown as { _data: unknown[] })._data = data;
    (node as unknown as { _declaration: Predicate | undefined })._declaration = declaration ?? node;
    (node as any)._binding = null;
    (node as any)._hashCodeCached = false;
    (node as any)._hashCode = 0;
    (node as any)._nrOfUnbound = -1;
    return node;
  }

  protected override struct(data: unknown[], declaration?: Predicate): Predicate {
    return ToInteger.fromDataToInteger(data, declaration ?? this.declaration());
  }

  // @JAVA_REF org.modelingvalue.nelumbo.strings.ToInteger#infer(int, InferContext)
  protected override inferInternal(_nrOfUnbound: number, _context: InferContext): InferResult {
    const nrOfUnbound = this.nrOfUnbound();
    if (nrOfUnbound > 1) {
      return this.unresolvable();
    }

    const integer = this.getIntegerVal(0);
    const string = this.getStringVal(1);

    if (string !== null) {
      try {
        const parsed = BigInt(parseInt(string, 10));
        if (integer !== null) {
          const eq = integer === parsed;
          return eq ? this.factCC() : this.falsehoodCC();
        } else {
          return (this.set(0, NInteger.of(parsed)) as ToInteger).factCI();
        }
      } catch {
        return integer !== null ? this.falsehoodCC() : this.falsehoodCI();
      }
    } else if (integer !== null) {
      const s = integer.toString();
      return (this.set(1, NString.of(s)) as ToInteger).factCI();
    }

    return this.unknown();
  }

  private getIntegerVal(index: number): bigint | null {
    const val = this.get(index);
    if (val instanceof NInteger) {
      return val.value();
    }
    return null;
  }

  private getStringVal(index: number): string | null {
    const val = this.get(index);
    if (val instanceof NString) {
      return val.value();
    }
    return null;
  }
}
